feat: advanced dimension filter

AdityaHegde opened the pull request:
closes #6710

- [x] Add comma separated "In List" search mode that is essentially an `in` filter.
- [x] Add a "Contains" search mode that is essentially a `like` filter.
- [x] Improve the way we differentiate select and bulk list since both are `in` query in the end.
- [x] Compress the url params using gzip https://github.com/rilldata/rill/pull/6748
- [ ] ~~Add tests for canvas filters.~~ Will be in a separate PR
- [x] UXQA 

**Checklist:**
- [x] Covered by tests
- [x] Ran it and it works as intended
- [x] Reviewed the diff before requesting a review
- [x] Checked for unhandled edge cases
- [x] Linked the issues it closes
- [ ] Checked if the docs need to be updated
- [x] Intend to cherry-pick into the release branch
- [x] I'm proud of this work!

ericpgreen2 commented:
Leaving our notes from our live discussion about a filter state refactor:
- This PR further motivates a refactor to our client-side filters state. It's now making less sense for our client-side mental model to be anchored on the `where: V1Expression` parameter that the backend expects. Rather, our mental model should mimic the user-facing UX, which now notably includes different filter modes. 
- Our dimension filter state should revolve around a type that makes `mode` explicit, e.g.:
    ```ts
   interface DimensionFilter {
      dimensionName: string;
      mode: "Select" | "Contains" | "In List";
      payload: string | string[]; // this is just for illustration because we should avoid this mixed type
      exclude: boolean;
    }
    ```
- Regardless of what the URL parameters are, we can deserialize them into the above structure (e.g. `const dimensionFilters = filters.fromURLParams(...)`). Regardless of what form our client-side mental model takes, we can serialize it into the `where` property that the metrics view queries expect (e.g. `const where = filters.toWhereExpression(...)`.

ericpgreen2 reviewed (changes requested):
Leaving various observations from my first pass

AdityaHegde commented on web-common/src/components/button/IconButton.svelte:
Hmmm not sure why this change showed up. It is not being used at all.

AdityaHegde commented on web-common/src/features/dashboards/filters/dimension-filters/DimensionFilter.svelte:
```diff
@@ -1,21 +1,37 @@
   export let label: string;
   export let selectedValues: string[];
+  export let searchText: string | undefined;
+  export let isMatchList: boolean | undefined;
```
We can reuse `DimensionFilterMode`

ericpgreen2 commented on web-common/src/features/dashboards/filters/Filters.svelte:
```diff
@@ -405,6 +409,10 @@
                 onToggleFilterMode={() => toggleDimensionFilterMode(name)}
                 onSelect={(value) =>
                   toggleDimensionValueSelection(name, value, true)}
+                onBulkSelect={(values) =>
+                  applyDimensionBulkSearch(name, values)}
```
We need to use consistent terminology. It's hard to keep track of "bulk select" vs "bulk search" vs  "in list".

ericpgreen2 commented on web-common/src/features/dashboards/filters/Filters.svelte:
```diff
@@ -405,6 +409,10 @@
                   toggleDimensionValueSelection(name, value, true)}
+                onBulkSelect={(values) =>
+                  applyDimensionBulkSearch(name, values)}
+                onSearch={(searchText) =>
+                  applyDimensionSearch(name, searchText)}
```
Similarly, this should more clearly map to the UI's "contains" terminology

ericpgreen2 reviewed (changes requested):
I haven't finished going through the whole PR yet, but here's some feedback for now.

UXQA:

1. The menu item hover state is mis-aligned with the above elements. 
    <img width="446" alt="image" src="https://github.com/user-attachments/assets/5e2bf326-54af-4cdd-a6af-f24cbabfb32b" />
2. Currently, when I submit the "Contains" filter with an empty input, all values are selected. Instead, there should be form validation that prevents the form from being submitted with an empty string. There should be a small error message in red text.
    <img width="1744" alt="image" src="https://github.com/user-attachments/assets/5cab4f43-8555-41ee-95f3-cd031742a7e6" />
3. Similar to 2, the InList mode needs form validation to prevent a filter from having an empty string input.
4. Why does the "Contains" filter match a different number of values in the dashboard (700) and the alert context (1018)?
    <img width="1184" alt="image" src="https://github.com/user-attachments/assets/16bd3552-d168-4c9e-9ee4-e05195f2e5a4" />

AdityaHegde commented on web-admin/src/features/alerts/metadata/AlertFilters.svelte:
```diff
@@ -47,8 +48,13 @@
   );
 
-  $: currentDimensionFilters = getDimensionFilters(dimensionIdMap, whereFilter);
+  $: currentDimensionFilters = getDimensionFilters(
+    dimensionIdMap,
+    whereFilter,
+    dimensionsWithInlistFilter,
+  );
   $: currentMeasureFilters = getMeasureFilters(measureIdMap, havingFilter);
```
You mean just for `dimensionsWithInlistFilter`. The other fields are taken from the query, those could be different.

AdityaHegde commented on web-admin/src/features/alerts/selectors.ts:
```diff
@@ -97,3 +102,40 @@ export function useIsAlertCreatedByCode(instanceId: string, name: string) {
   );
 }
+
+export function useAlertDashboardState(
+  instanceId: string,
+  alertSpec: V1AlertSpec | undefined,
+) {
```
Ya that is one option. It would lead to `.subscribe(set)` pattern so I think it is better this way. Besides both consumers of this already have `alertSpec`

AdityaHegde commented on web-common/src/components/tag/Tag.svelte:
```diff
@@ -15,6 +15,7 @@
   export let color: Color = "gray";
   export let height = 21;
   export let text: string = "";
+  export let noSpan = false;
```
Seems like another merge issue

AdityaHegde commented on web-common/src/features/canvas/stores/filters.ts:
```diff
@@ -556,10 +623,12 @@ export class Filters {
-    if (
-      expr?.cond?.op !== V1Operation.OPERATION_AND &&
-      expr?.cond?.op !== V1Operation.OPERATION_OR
+    let { expr } = convertFilterParamToExpression(filterText); // TODO: use dimensionsWithInlistFilter
```
yes, good catch

AdityaHegde commented:
Thanks @ericpgreen2 

For 1, I see its the case on prod as well. Lets address it as a bug fix in a follow up.
Addressed 2 and 3 but disabling apply. There is no design for error message. Since we disable apply it should be clear to the user. Lets take the improvement in a follow up.
Fixed 4 by using the correct time ranges.

ericpgreen2 commented on web-common/src/features/dashboards/state-managers/selectors/dimension-filters.ts:
```diff
@@ -21,22 +25,86 @@ export const selectedDimensionValues = (
+    });
+  }
+
+  return readable({
+    isFetching: false,
+    isLoading: false,
+    error: undefined,
+    data: [],
+  });
```
When will this get hit? Should this return an error?

AdityaHegde replied:
It shouldnt be hit. Adding an error to make the code readable.

ericpgreen2 commented on web-common/src/features/dashboards/state-managers/selectors/dimension-filters.ts:
```diff
@@ -21,22 +25,86 @@ export const selectedDimensionValues = (
+  const dimExpr = whereFilter.cond?.exprs?.find((e) =>
+    matchExpressionByName(e, dimensionName),
+  );
+  if (!dimExpr?.cond?.op)
+    return readable({
+      isFetching: false,
+      isLoading: false,
+      error: undefined,
+      data: [],
+    });
```
Is this a type guard? Or should this return an error?

AdityaHegde replied:
This is the case when a dimension has no filter. It should return empty array similar to how it was before.

ericpgreen2 commented on web-common/src/features/dashboards/state-managers/selectors/dimension-filters.ts:
```diff
@@ -21,22 +25,86 @@ export const selectedDimensionValues = (
+    return readable({
+      isFetching: false,
+      isLoading: false,
+      error: undefined,
```
Should there be an error here, or no?

AdityaHegde replied:
No, this will break leaderboards when there is a complex filter present or no filter is present. Added a comment.

ericpgreen2 commented on web-common/src/features/dashboards/state-managers/actions/dimension-filters.ts:
```diff
@@ -63,6 +103,55 @@ export function toggleDimensionValueSelection(
+
+  if (!dashboard.whereFilter.cond?.exprs) return;
+
+  const isInclude = !dashboard.dimensionFilterExcludeMode.get(dimensionName);
+  const expr = createInExpression(dimensionName, values, !isInclude);
```
Here there's still a double negation. Why not use `isExclude`?

ericpgreen2 commented on web-common/src/features/dashboards/state-managers/actions/dimension-filters.ts:
```diff
@@ -63,6 +103,55 @@ export function toggleDimensionValueSelection(
+
+  if (!dashboard.whereFilter.cond?.exprs) return;
+
+  const isInclude = !dashboard.dimensionFilterExcludeMode.get(dimensionName);
+  const expr = createLikeExpression(
+    dimensionName,
+    `%${searchText}%`,
+    !isInclude,
+  );
```
Double negation

ericpgreen2 commented on web-common/src/features/dashboards/filters/dimension-filters/DimensionFilter.spec.ts:
```diff
@@ -0,0 +1,252 @@
+    );
+  });
+
+  it("Select mode filter", async () => {
```
This is a bit hard to follow. At first, given the name of the test. I was confused when I saw "Contains" logic. It'd help if there were sectional comments at the _beginning_ of the "Contains" and "In list" parts of the test, not just at the end where those assertion happens.

ericpgreen2 replied:
Sectional comments would help within each of this file's tests

ericpgreen2 commented on web-common/src/features/dashboards/filters/dimension-filters/DimensionFilter.spec.ts:
```diff
@@ -0,0 +1,252 @@
+    );
+  });
+
+  it("Search mode filter", async () => {
```
Outdated terminology – should be "contains"

ericpgreen2 commented on web-common/src/features/dashboards/filters/dimension-filters/DimensionFilter.spec.ts:
```diff
@@ -0,0 +1,252 @@
+    );
+  });
+
+  it("Bulk mode filter using dropdown", async () => {
```
Outdated terminology – should be "in list"

ericpgreen2 commented on web-common/src/features/dashboards/filters/dimension-filters/DimensionFilter.spec.ts:
```diff
@@ -0,0 +1,252 @@
+    );
+  });
+
+  it("Bulk mode filter using search text", async () => {
```
Outdated terminology

ericpgreen2 commented on web-common/src/features/dashboards/filters/dimension-filters/dimension-filter-values.ts:
Can you please add function-level comments to this file? It'd help to more easily understand what they're doing.

ericpgreen2 commented on web-common/src/features/dashboards/filters/dimension-filters/DimensionFilter.svelte:
```diff
@@ -24,54 +40,237 @@
+
+  $: checkSearchText(curSearchText);
+
+  let searchedBulkValues: string[] =
```
Nit: Can you please add some sensible line breaks from lines 60 to 114? Without any line breaks, it's pretty overwhelming.

AdityaHegde commented on web-admin/src/features/public-urls/CreatePublicURLForm.svelte:
```diff
@@ -242,9 +245,12 @@
               filters={$dashboardStore.whereFilter}
+              dimensionsWithInlistFilter={$dashboardStore.dimensionsWithInlistFilter}
               dimensionThresholdFilters={$dashboardStore.dimensionThresholdFilters}
               timeRange={undefined}
               comparisonTimeRange={undefined}
+              resolvedTimeStart={$timeControlStore.timeStart}
+              resolvedTimeEnd={$timeControlStore.timeEnd}
```
Renaming to `queryTime{Start,End}` and `displayTimeRange` to make it self-explanatory.

AdityaHegde commented on web-common/src/components/chip/removable-list-chip/RemovableListBody.svelte:
Hmmm it is only used for dimension filter. It is probably simpler to just move it there.

AdityaHegde commented on web-local/tests/explores/explores.spec.ts:
Hmm the `publisher` come from the `name` of the dimension. Lets do this standardisation as our testing improvement effort.

AdityaHegde commented on web-common/src/features/dashboards/leaderboard/Leaderboard.svelte:
```diff
@@ -53,7 +54,7 @@
   export let timeRange: V1TimeRange;
   export let comparisonTimeRange: V1TimeRange | undefined;
-  export let selectedValues: string[];
+  export let selectedValues: CompoundQueryResult<string[]>;
```
Hmmm this is not so trivial refactor. Since the Leaderboard component is called within a loop we cant just use `$` to get the values. Variables defined in a const statement cannot be used with a `$` either. Lets do this in a follow up.

ericpgreen2 reviewed (approved):
Nice work. I love the new "Contains" mode!

Given this is a huge PR, IMO cherry-picking it onto the release branch to be released in ~36 hours is a bit aggressive. Unless any of @mindspank, @royendo, @jkhwu, @ericokuma have time to sufficiently kick the tires.
